File: lib_js/models/supabase/organization.js
import { SupabaseModel } from "./supabaseModel.js";
import { ACLGroupUsersModel } from "./acl.js"; // Import ACLGroupUsersModel

// Define UserProfileModel
export class UserProfileModel extends SupabaseModel {
    static TABLE_NAME = "user_profile";
    static TABLE_FIELDS = {
        id: { type: "uuid", required: false, dbColumn: "id" },

        authId: { type: "uuid", required: false, dbColumn: "auth_id" },
        email: { type: "string", required: false, dbColumn: "email" },
        name: { type: "string", required: false, dbColumn: "name" },
        profilePicture: {
            type: "string",
            required: false,
            dbColumn: "profile_picture"
        },
        metadata: { type: "json", required: false, dbColumn: "metadata" },
        lang: { type: "string", required: false, dbColumn: "lang" },
        activePanelId: {
            type: "uuid",
            required: false,
            dbColumn: "active_panel_id"
        },
        preferences: { type: "json", required: false, dbColumn: "preferences" },
        paymentDetails: {
            type: "json",
            required: false,
            dbColumn: "payment_details"
        },
        disabled: { type: "boolean", required: false, dbColumn: "disabled" },
        disabledAt: { type: "date", required: false, dbColumn: "disabled_at" },
        createdAt: { type: "date", required: false, dbColumn: "created_at" },
        updatedAt: { type: "date", required: false, dbColumn: "updated_at" },
        activeOrganizationId: {
            type: "uuid",
            required: false,
            dbColumn: "active_organization_id"
        },
        activeConversationId: {
            type: "uuid",
            required: false,
            dbColumn: "active_conversation_id"
        }
    };
}

// Define UserDataModel
export class UserDataModel extends SupabaseModel {
    static TABLE_NAME = "user_data";
    static TABLE_FIELDS = {
        id: { type: "uuid", required: false, dbColumn: "id" },
        authId: { type: "uuid", required: false, dbColumn: "auth_id" },
        item: { type: "string", required: true, dbColumn: "item" },
        targetId: { type: "uuid", required: false, dbColumn: "target_id" },
        data: { type: "json", required: false, dbColumn: "data" },
        createdAt: { type: "date", required: false, dbColumn: "created_at" },
        updatedAt: { type: "date", required: false, dbColumn: "updated_at" }
    };
}

// Define OrganizationRoleModel
export class OrganizationRoleModel extends SupabaseModel {
    static TABLE_NAME = "organization_role";
    static TABLE_FIELDS = {
        id: { type: "uuid", required: false, dbColumn: "id" },
        name: { type: "string", required: true, dbColumn: "name" },
        description: {
            type: "string",
            required: false,
            dbColumn: "description"
        },
        seniority: { type: "number", required: true, dbColumn: "seniority" },
        metadata: { type: "json", required: false, dbColumn: "metadata" },
        disabled: { type: "boolean", required: false, dbColumn: "disabled" },
        disabledAt: { type: "date", required: false, dbColumn: "disabled_at" },
        createdAt: { type: "date", required: false, dbColumn: "created_at" },
        updatedAt: { type: "date", required: false, dbColumn: "updated_at" },
        organizationId: {
            type: "uuid",
            required: false,
            dbColumn: "organization_id"
        }
    };
}

// Define OrganizationTeamModel
export class OrganizationTeamModel extends SupabaseModel {
    static TABLE_NAME = "organization_team";
    static TABLE_FIELDS = {
        id: { type: "uuid", required: false, dbColumn: "id" },
        name: { type: "string", required: true, dbColumn: "name" },
        metadata: { type: "json", required: false, dbColumn: "metadata" },
        disabled: { type: "boolean", required: false, dbColumn: "disabled" },
        disabledAt: { type: "date", required: false, dbColumn: "disabled_at" },
        createdAt: { type: "date", required: false, dbColumn: "created_at" },
        updatedAt: { type: "date", required: false, dbColumn: "updated_at" },
        ownerId: { type: "uuid", required: false, dbColumn: "owner_id" },
        organizationId: {
            type: "uuid",
            required: false,
            dbColumn: "organization_id"
        }
    };
}

// Define OrganizationTeamMembersModel
export class OrganizationTeamMembersModel extends SupabaseModel {
    static TABLE_NAME = "organization_team_members";
    static TABLE_FIELDS = {
        authId: { type: "uuid", required: true, dbColumn: "auth_id" },
        userId: { type: "uuid", required: true, dbColumn: "user_id" },
        teamId: { type: "uuid", required: true, dbColumn: "team_id" },
        roleId: { type: "uuid", required: true, dbColumn: "role_id" },
        disabled: { type: "boolean", required: false, dbColumn: "disabled" },
        disabledAt: { type: "date", required: false, dbColumn: "disabled_at" },
        createdAt: { type: "date", required: false, dbColumn: "created_at" },
        updatedAt: { type: "date", required: false, dbColumn: "updated_at" },
        organizationId: {
            type: "uuid",
            required: false,
            dbColumn: "organization_id"
        }
    };

    static async saveToSupabase(
        supabase,
        instance,
        onConflict = ["authId", "teamId"]
    ) {
        return super.saveToSupabase(supabase, instance, onConflict);
    }

    static async upsertToSupabase(
        supabase,
        instances,
        onConflict = ["authId", "teamId"],
        idColumn = null
    ) {
        return super.upsertToSupabase(
            supabase,
            instances,
            onConflict,
            idColumn
        );
    }

    static async fetchFromSupabase(supabase, value = null, idColumn = null) {
        if (value instanceof this) {
            value = {
                authId: value.attributes.authId,
                teamId: value.attributes.teamId
            };
        }
        return super.fetchFromSupabase(supabase, value, idColumn);
    }

    static async existsInSupabase(supabase, value = null, idColumn = null) {
        if (value instanceof this) {
            value = {
                authId: value.attributes.authId,
                teamId: value.attributes.teamId
            };
        }
        return super.existsInSupabase(supabase, value, idColumn);
    }

    static async deleteFromSupabase(supabase, value = null, idColumn = null) {
        if (value instanceof this) {
            value = {
                authId: value.attributes.authId,
                teamId: value.attributes.teamId
            };
        }
        return super.deleteFromSupabase(supabase, value, idColumn);
    }
}

// Define OrganizationUsersModel
export class OrganizationUsersModel extends SupabaseModel {
    static TABLE_NAME = "organization_users";
    static TABLE_FIELDS = {
        authId: { type: "uuid", required: false, dbColumn: "auth_id" },
        userId: { type: "uuid", required: false, dbColumn: "user_id" },
        organizationId: {
            type: "uuid",
            required: false,
            dbColumn: "organization_id"
        },
        metadata: { type: "json", required: false, dbColumn: "metadata" },
        isAdmin: { type: "boolean", required: false, dbColumn: "is_admin" },
        disabled: { type: "boolean", required: false, dbColumn: "disabled" },
        disabledAt: { type: "date", required: false, dbColumn: "disabled_at" },
        createdAt: { type: "date", required: false, dbColumn: "created_at" },
        updatedAt: { type: "date", required: false, dbColumn: "updated_at" }
    };

    static async saveToSupabase(
        supabase,
        instance,
        onConflict = ["authId", "organizationId"]
    ) {
        return super.saveToSupabase(supabase, instance, onConflict);
    }

    static async upsertToSupabase(
        supabase,
        instances,
        onConflict = ["authId", "organizationId"],
        idColumn = null
    ) {
        return super.upsertToSupabase(
            supabase,
            instances,
            onConflict,
            idColumn
        );
    }

    static async fetchFromSupabase(supabase, value = null, idColumn = null) {
        if (value instanceof this) {
            value = {
                authId: value.attributes.authId,
                organizationId: value.attributes.organizationId
            };
        }
        return super.fetchFromSupabase(supabase, value, idColumn);
    }

    static async existsInSupabase(supabase, value = null, idColumn = null) {
        if (value instanceof this) {
            value = {
                authId: value.attributes.authId,
                organizationId: value.attributes.organizationId
            };
        }
        return super.existsInSupabase(supabase, value, idColumn);
    }

    static async deleteFromSupabase(supabase, value = null, idColumn = null) {
        if (value instanceof this) {
            value = {
                authId: value.attributes.authId,
                organizationId: value.attributes.organizationId
            };
        }
        return super.deleteFromSupabase(supabase, value, idColumn);
    }

    async inAclGroup(supabase, aclGroupId) {
        return await ACLGroupUsersModel.existsInSupabase(supabase, {
            userId: this.attributes.userId,
            aclGroupId
        });
    }

    async connectWithAclGroup(supabase, aclGroupId, acl) {
        // Attempt to fetch the existing relationship
        const existingAclGroupUser = await ACLGroupUsersModel.fetchFromSupabase(
            supabase,
            { userId: this.attributes.userId, aclGroupId }
        );

        if (existingAclGroupUser) {
            // If the relationship exists, update the ACL level
            existingAclGroupUser.attributes.acl = acl;
            await existingAclGroupUser.update();
        } else {
            // Otherwise, create a new ACL group user relationship
            const aclGroupUser = new ACLGroupUsersModel({
                userId: this.attributes.userId,
                authId: this.attributes.authId,
                aclGroupId,
                acl,
                organizationId: this.attributes.organizationId
            });
            await ACLGroupUsersModel.saveToSupabase(supabase, aclGroupUser);
        }
    }

    async disconnectWithAclGroup(supabase, aclGroupId) {
        await ACLGroupUsersModel.deleteFromSupabase(supabase, {
            userId: this.attributes.userId,
            aclGroupId
        });
    }

    async create(supabase, groupsWithLevels = null) {
        const orgUser = await this.saveToSupabase(supabase, this);

        if (groupsWithLevels) {
            for (const [aclGroupId, acl] of Object.entries(groupsWithLevels)) {
                await orgUser.connectWithAclGroup(supabase, aclGroupId, acl);
            }
        }

        return orgUser;
    }
}

// Define OrganizationsModel
export class OrganizationsModel extends SupabaseModel {
    static TABLE_NAME = "organizations";
    static TABLE_FIELDS = {
        id: { type: "uuid", required: false, dbColumn: "id" },
        defaultAclGroupId: {
            type: "uuid",
            required: false,
            dbColumn: "default_acl_group_id"
        },
        defaultBucketId: {
            type: "string",
            required: false,
            dbColumn: "default_bucket_id"
        },
        name: { type: "string", required: false, dbColumn: "name" },
        website: { type: "string", required: false, dbColumn: "website" },
        logo: { type: "string", required: false, dbColumn: "logo" },
        metadata: { type: "json", required: false, dbColumn: "metadata" },
        disabled: { type: "boolean", required: false, dbColumn: "disabled" },
        disabledAt: { type: "date", required: false, dbColumn: "disabled_at" },
        createdAt: { type: "date", required: false, dbColumn: "created_at" },
        updatedAt: { type: "date", required: false, dbColumn: "updated_at" },
        ownerId: { type: "uuid", required: false, dbColumn: "owner_id" }
    };
}

File: source/models/supabase/organization.py
import json
from typing import ClassVar, Dict, Optional
from uuid import UUID
from pydantic import Field, field_validator
from pydantic.types import PositiveInt
from datetime import datetime
from supabase.client import AsyncClient

from source.models.supabase.supabase_model import SupabaseModel
from source.models.supabase.acl import ACLGroupUsersModel, UserACL


class UserProfileModel(SupabaseModel):
    TABLE_NAME: ClassVar[str] = "user_profile"
    id: Optional[UUID] = Field(default=None)
    auth_id: Optional[UUID] = Field(default=None)
    email: Optional[str] = Field(default=None)
    name: Optional[str] = Field(default=None)
    profile_picture: Optional[str] = Field(default=None)
    metadata: Optional[Dict] = Field(default=None)
    lang: Optional[str] = Field(default=None)
    active_panel_id: Optional[UUID] = Field(default=None)
    preferences: Optional[Dict] = Field(default=None)
    payment_details: Optional[Dict] = Field(default=None)
    disabled: bool = Field(default=False)
    disabled_at: Optional[datetime] = Field(default=None)
    created_at: Optional[datetime] = Field(default=None)
    updated_at: Optional[datetime] = Field(default=None)
    active_organization_id: Optional[UUID] = Field(default=None)
    active_conversation_id: Optional[UUID] = Field(default=None)

    @field_validator("metadata", mode="before")
    def validate_metadata(cls, v):
        if isinstance(v, str):
            return json.loads(v)
        elif isinstance(v, dict):
            return json.dumps(v)
        return v


class UserDataModel(SupabaseModel):
    TABLE_NAME: ClassVar[str] = "user_data"
    id: Optional[UUID] = Field(default=None)
    auth_id: Optional[UUID] = Field(default=None)
    item: str
    target_id: Optional[UUID] = Field(default=None)
    data: Optional[Dict] = Field(default=None)
    created_at: Optional[datetime] = Field(default=None)
    updated_at: Optional[datetime] = Field(default=None)


class OrganizationRoleModel(SupabaseModel):
    TABLE_NAME: ClassVar[str] = "organization_role"
    id: Optional[UUID] = Field(default=None)
    name: str
    description: Optional[str] = Field(default=None)
    seniority: PositiveInt
    metadata: Optional[Dict] = Field(default=None)
    disabled: bool = Field(default=False)
    disabled_at: Optional[datetime] = Field(default=None)
    created_at: Optional[datetime] = Field(default=None)
    updated_at: Optional[datetime] = Field(default=None)
    organization_id: Optional[UUID] = Field(default=None)

    @field_validator("metadata", mode="before")
    def validate_metadata(cls, v):
        if isinstance(v, str):
            return json.loads(v)
        elif isinstance(v, dict):
            return json.dumps(v)
        return v


class OrganizationTeamModel(SupabaseModel):
    TABLE_NAME: ClassVar[str] = "organization_team"
    id: Optional[UUID] = Field(default=None)
    name: str
    metadata: Optional[Dict] = Field(default=None)
    disabled: bool = Field(default=False)
    disabled_at: Optional[datetime] = Field(default=None)
    created_at: Optional[datetime] = Field(default=None)
    updated_at: Optional[datetime] = Field(default=None)
    owner_id: Optional[UUID] = Field(default=None)
    organization_id: Optional[UUID] = Field(default=None)

    @field_validator("metadata", mode="before")
    def validate_metadata(cls, v):
        if isinstance(v, str):
            return json.loads(v)
        elif isinstance(v, dict):
            return json.dumps(v)
        return v


class OrganizationTeamMembersModel(SupabaseModel):
    TABLE_NAME: ClassVar[str] = "organization_team_members"
    auth_id: UUID
    user_id: UUID
    team_id: UUID
    role_id: UUID
    disabled: bool = Field(default=False)
    disabled_at: Optional[datetime] = Field(default=None)
    created_at: Optional[datetime] = Field(default=None)
    updated_at: Optional[datetime] = Field(default=None)
    organization_id: Optional[UUID] = Field(default=None)

    @classmethod
    async def save_to_supabase(
        cls, supabase: AsyncClient, instance, on_conflict=["auth_id", "team_id"]
    ):
        return await super(OrganizationTeamMembersModel, cls).save_to_supabase(
            supabase, instance, on_conflict
        )

    @classmethod
    async def upsert_to_supabase(
        cls,
        supabase: AsyncClient,
        instances,
        on_conflict=["auth_id", "team_id"],
        id_column=None,
    ):
        return await super(OrganizationTeamMembersModel, cls).upsert_to_supabase(
            supabase, instances, on_conflict, id_column
        )

    @classmethod
    async def fetch_from_supabase(
        cls, supabase: AsyncClient, value=None, id_column=None
    ):
        if isinstance(value, cls):
            value = {"auth_id": value.auth_id, "team_id": value.team_id}
        return await super(OrganizationTeamMembersModel, cls).fetch_from_supabase(
            supabase, value=value, id_column=id_column
        )

    @classmethod
    async def exists_in_supabase(
        cls, supabase: AsyncClient, value=None, id_column=None
    ):
        if isinstance(value, cls):
            value = {"auth_id": value.auth_id, "team_id": value.team_id}
        return await super(OrganizationTeamMembersModel, cls).exists_in_supabase(
            supabase, value=value, id_column=id_column
        )

    @classmethod
    async def delete_from_supabase(
        cls, supabase: AsyncClient, value=None, id_column=None
    ):
        if isinstance(value, cls):
            value = {"auth_id": value.auth_id, "team_id": value.team_id}
        return await super(OrganizationTeamMembersModel, cls).delete_from_supabase(
            supabase, value=value, id_column=id_column
        )


class OrganizationUsersModel(SupabaseModel):
    TABLE_NAME: ClassVar[str] = "organization_users"
    auth_id: Optional[UUID] = Field(default=None)
    user_id: Optional[UUID] = Field(default=None)
    organization_id: Optional[UUID] = Field(default=None)
    metadata: Optional[Dict] = Field(default=None)
    is_admin: bool = Field(default=False)
    disabled: bool = Field(default=False)
    disabled_at: Optional[datetime] = Field(default=None)
    created_at: Optional[datetime] = Field(default=None)
    updated_at: Optional[datetime] = Field(default=None)

    @field_validator("metadata", mode="before")
    def validate_metadata(cls, v):
        if isinstance(v, str):
            return json.loads(v)
        elif isinstance(v, dict):
            return json.dumps(v)
        return v

    @classmethod
    async def save_to_supabase(
        cls, supabase: AsyncClient, instance, on_conflict=["auth_id", "organization_id"]
    ):
        return await super(OrganizationUsersModel, cls).save_to_supabase(
            supabase, instance, on_conflict
        )

    @classmethod
    async def upsert_to_supabase(
        cls,
        supabase: AsyncClient,
        instances,
        on_conflict=["auth_id", "organization_id"],
        id_column=None,
    ):
        return await super(OrganizationUsersModel, cls).upsert_to_supabase(
            supabase, instances, on_conflict, id_column
        )

    @classmethod
    async def fetch_from_supabase(
        cls, supabase: AsyncClient, value=None, id_column=None
    ):
        if isinstance(value, cls):
            value = {"auth_id": value.auth_id, "organization_id": value.organization_id}
        return await super(OrganizationUsersModel, cls).fetch_from_supabase(
            supabase, value=value, id_column=id_column
        )

    @classmethod
    async def exists_in_supabase(
        cls, supabase: AsyncClient, value=None, id_column=None
    ):
        if isinstance(value, cls):
            value = {"auth_id": value.auth_id, "organization_id": value.organization_id}
        return await super(OrganizationUsersModel, cls).exists_in_supabase(
            supabase, value=value, id_column=id_column
        )

    @classmethod
    async def delete_from_supabase(
        cls, supabase: AsyncClient, value=None, id_column=None
    ):
        if isinstance(value, cls):
            value = {"auth_id": value.auth_id, "organization_id": value.organization_id}
        return await super(OrganizationUsersModel, cls).delete_from_supabase(
            supabase, value=value, id_column=id_column
        )

    async def in_acl_group(self, supabase: AsyncClient, acl_group_id: UUID):
        return await ACLGroupUsersModel.exists_in_supabase(
            supabase, value={"user_id": self.user_id, "acl_group_id": acl_group_id}
        )

    async def connect_with_acl_group(
        self, supabase: AsyncClient, acl_group_id: UUID, acl: UserACL
    ):
        """Connect the user to an ACL group with specified ACL level."""
        # Attempt to fetch the existing relationship
        existing_acl_group_user = await ACLGroupUsersModel.fetch_from_supabase(
            supabase, {"user_id": self.user_id, "acl_group_id": acl_group_id}
        )

        if existing_acl_group_user:
            # If the relationship exists, update the ACL level
            existing_acl_group_user.acl = acl
            await existing_acl_group_user.update()
        else:
            # Otherwise, create a new ACL group user relationship
            acl_group_user = ACLGroupUsersModel(
                user_id=self.user_id,
                auth_id=self.auth_id,
                acl_group_id=acl_group_id,
                acl=acl,
                organization_id=self.organization_id,
            )
            await ACLGroupUsersModel.save_to_supabase(supabase, acl_group_user)

    async def disconnect_with_acl_group(
        self, supabase: AsyncClient, acl_group_id: UUID
    ):
        """Disconnect the user from an ACL group."""
        await ACLGroupUsersModel.delete_from_supabase(
            supabase, {"user_id": self.user_id, "acl_group_id": acl_group_id}
        )

    async def create(
        self,
        supabase: AsyncClient,
        groups_with_levels: dict[UUID, UserACL] = None,
    ) -> "OrganizationUsersModel":
        """Create a new organization user and connect to specified ACL groups."""
        # Save the organization user
        org_user = await self.save_to_supabase(supabase, self)

        # Connect to specified ACL groups
        if groups_with_levels:
            for acl_group_id, acl in groups_with_levels.items():
                await org_user.connect_with_acl_group(supabase, acl_group_id, acl)

        return org_user


class OrganizationsModel(SupabaseModel):
    TABLE_NAME: ClassVar[str] = "organizations"
    id: Optional[UUID] = Field(default=None)
    default_acl_group_id: Optional[UUID] = Field(default=None)
    default_bucket_id: Optional[str] = Field(default=None)
    name: Optional[str] = Field(default=None)
    website: Optional[str] = Field(default=None)
    logo: Optional[str] = Field(default=None)
    metadata: Optional[Dict] = Field(default=None)
    disabled: bool = Field(default=False)
    disabled_at: Optional[datetime] = Field(default=None)
    created_at: Optional[datetime] = Field(default=None)
    updated_at: Optional[datetime] = Field(default=None)
    owner_id: Optional[UUID] = Field(default=None)

    @field_validator("metadata", mode="before")
    def validate_metadata(cls, v):
        if isinstance(v, str):
            return json.loads(v)
        elif isinstance(v, dict):
            return json.dumps(v)
        return v
